### PR TITLE
Resolves #1824: Use different analyzers for different fields in one L…

### LINKED ISSUE #1824: Use different analyzers for different fields in one Lucene index, and enable exclusion of fields for auto-complete

### PR DESCRIPTION
**Pull request by tian-yizuo**
…ucene index, and enable exclusion of fields for auto-complete

### CODE REVIEW

**Review by MMcM — APPROVED** [review]
Although I think the scheme discussed in the meeting of having per-field option _names_ with a common prefix (followed by the field name) would be more straightforward, it looks like this ought to work okay.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexOptions.java
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexOptions.java:75]:
```suggestion
     * The format of the value should be "fieldName1/analyzerName1/fieldName2/analyzerName2...", with keys and values split by "/".
```

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexOptions.java
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexOptions.java:85]:
```suggestion
     * The format of the value should be "fieldName1/fieldName2/...", with field names split by "/".
```

**Review by alecgrieser — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerCombinationProvider.java
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerCombinationProvider.java:39]: The ```AnalyzerChooser``` is designed for the provision of an Lucene analyzer given a text, so it is per Lucene analyzer. The only reason we need it for now, is for the capability needed to use different one for different languages. Whereas the ```LuceneAnalyzerCombinationProvider``` is more like a single object to provide a pair of analyzers for index time and query time based on the options config, because we want to make is simple for clients to configure for the analyzers to use, without worrying about which to use on index time and which to use on query time.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java:103]: @alecgrieser yeah it sounds better to use different delineators for that between key-value pairs and that between keys and values. So can I assume no field names will have ```:``` or ```,```?

Comment (alecgrieser) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java:103]: I don't know how valid an assumption that is; I think `:` and `,` are better delineator characters for maps, but I don't know whether they're currently allowed in field names (or analyzer names) or not. That's one of the reasons that for, say, the "replaced by" index name options, rather than coming up with a scheme by which the index names could be put together in a single option, we went with instead a single option _prefix_, and then collected all IndexOptions that began with that prefix to produce the list.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerCombinationProvider.java
Comment (alecgrieser) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerCombinationProvider.java:39]: I think I like the abstraction, for the most part, and that it makes sense to encapsulate index-and-query-time-analyzers into an object other than a pair. However, once you have that abstraction, I'm not sure it makes sense to combine that with the behavior that constructs the per-field Lucene analyzers.

I also think the abstraction is fine for now, though, and something we could iterate on later if we really want to

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:74]: This isn't exactly performance critical code, but this recomputes `keyValue.indexOf(':')` multiple times. Ideally, this would compute the value once (for each `keyValue`) and store it in a local variable.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:75]:
```suggestion
                    throw new MetaDataException("Index " + index.getName() + " has invalid option value " + optionKey + ": " + analyzerNamePerFieldOption);
```

Likewise for the other `MetaDataExceptions` that are thrown in this class

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:69]: I think these checks are all extraneous, as they'd all get validated by the later check of each key-value pair.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:63]: I don't think the way that `strip()` is handled here is quite right, because `strip` is not called when the index options are actually evaluated later

**Review by alecgrieser — CHANGES_REQUESTED** [review]
In my original review, I also proposed stripping out whitespace from the field and analyzer name options, but that wasn't done here, except somewhat during validation. I think it would be reasonable to either go with my initial suggestion (and strip out whitespace for each field and analyzer name) or to have tighter validation that required that there not be whitespace between elements.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:63]: Changed it to also call the ```strip()``` when actually evaluating the index options.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java]: Good catch, removed these

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java]: Fixed it

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java]: Fixed it

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:36]: I'm not sure this belongs in the `LuceneIndexValidatorTest`, but we probably want positive test cases somewhere as well. I...suppose this doesn't need to block this PR, though

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java:109]: As written, this parser algorithm would turn the following string:

```
   a:b,  c: d,e :f
```

Into:

```
{"a": "b", "c": " d", "e ", "f"}
```

That is, it strips out whitespace at the beginning of each key and the end of each value, but not at the end of each key and the beginning of each value. Is that intentional?

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:38]: It might also make sense to add test cases where there is whitespace, and then validate that the whitespace is stripped out as expected

**Review by alecgrieser — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:36]: The 1st assertion is just a positive test case.

Comment (alecgrieser) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:36]: Right, but I meant more, like, tests that validate that, for example, whitespace is handled correctly and that the strings are parsed into the expected sets or maps, etc.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java:109]: Fixed it

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:75]: Should this validate that the (stripped) key and value are non-empty?

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:42]: This should probably should be in an `assertDoesNotThrow` assertion to underscore why this is calling `validateIndexOptions`

**Review by alecgrieser — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java]: I think empty key or value should cause exception because the check ```keyValue.isEmpty() || firstIndex < 1 || firstIndex > keyValue.length() - 2```

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:39]: done

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java]: Done

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java
Comment (alecgrieser) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java]: Oh, I see. It's a little complicated, but I think you're right: when the `keyValue` is stripped, if there's a completely blank key or value, then either `firstIndex` will be < 1 (for a blank key) or `firstIndex` will be > length - 2 (for a blank value), assuming there's only one delimiter. But I had to convince myself that was enough

**Review by alecgrieser — APPROVED** [review]
Approved, pending a successful PRB (which it seems like is a style thing)

**Review by alecgrieser — APPROVED** [review]

**Review by MMcM — APPROVED** [review]

**Pull request merged**
